### PR TITLE
Deprecate torch.solve

### PR DESCRIPTION
**Pull request by lezcano**
Deprecate deprecate deprecate.

**BC-breaking Note:**

This PR deprecates torch.solve. It includes an upgrade guide in the documentation for torch.solve describing how to use torch.linalg.solve or torch.lu (or both) instead.

It DOES NOT remove torch.solve.

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -2618,9 +2618,8 @@ Tensor linalg_tensorsolve(const Tensor& self, const Tensor& other, optional<IntA
 
   self_ = self_.reshape({result_product, result_product});
 
-  // 0th output of at::solve is the solution
-  // normally `other` would be flattened by at::solve expects 2D input
-  Tensor result = std::get<0>(at::solve(other.reshape({other.numel(), 1}), self_));
+  // normally `other` would be flattened by at::linalg_solve expects 2D input
+  Tensor result = at::linalg_solve(self_, other.reshape({other.numel(), 1}));
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/LinearAlgebra.cpp:2622]:
```suggestion
  Tensor result = at::linalg_solve(self_, other.flatten());
```

**Review by IvanYashchuk — APPROVED** [review]

**Comment by Lezcano** [comment]
@mruberry this should be ready as well

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -733,6 +733,14 @@ std::tuple<Tensor, Tensor> _solve_helper_cpu(const Tensor& self, const Tensor& A
 
 // Supports arbitrary batch dimensions for self and A
 std::tuple<Tensor,Tensor> solve(const Tensor& self, const Tensor& A) {
+  TORCH_WARN_ONCE(
+    "torch.solve is deprecated in favor of torch.linalg.solve",
+    "and will be removed in a future PyTorch release.\n",
+    "torch.linalg.solve has its arguments reversed and does not return the LU factorization.\n",
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:739]: What should users do if they want the LU factorization?

**Review by mruberry — COMMENTED** [review]
Just two tweaks:

- upgrade guide needs a tweak to explain how to acquire the LU factorization, if desired
- fbgemm updates accidentally crept into the PR

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (mruberry) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:739]: I agree with you, but should we point people to lu and lu_solve here?

Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:739]: Would not hurt, I guess :)

**Comment by Lezcano** [comment]
Both points have been addressed. This should be ready to go @mruberry

**Review by mruberry — APPROVED** [review]
Cool! I think this is the last unmerged deprecation

File: torch/_torch_docs.py
Comment (ez2rok) [inline torch/_torch_docs.py:1]: It is important to note that the order of the arguments here is reversed.

Comment (lezcano) [reply torch/_torch_docs.py:1]: This is described in the `.. warning::` that we added to `torch.solve` below in this PR

**Pull request closed without merging**